keps/sig-api-machinery/20190723-coordinated-discovery-document.md

```diff
@@ -0,0 +1,181 @@
+   APIServerID string
+
+   // A hash calculated based on the discovery documents of built-in APIs. If
+   // two hashes equal, it means that the two API servers serve the same set of
+   // APIs.
+   // Required.
+   BuiltInAPIHash string
+
+   // LastHeartbeatTime is the last time the reporting API server updates this 
+   // field.
+   // Required 
+   LastHeartbeatTime metav1.Time
+}
+```
+
+API server instances report the hashes of their discovery documents through this
+API.  Although the main users of this API are the API servers, we make this API
+public to make debugging convenient.
+
+Another API change is that at each level of the discovery [API][] (APIGroupList,
+APIVersions, and APIResourceList), we add a new field `Consistent` to indicate
+if all API servers agree on the discovery documents, i.e., if all hashes
+reported in `DiscoveryDocHashes` are equal.
+
+Alternative to adding the `Consistent` field, we have considered letting API
+servers respond 500 error to discovery requests if there are multiple hash
+values in `DiscoveryDocHashes`. This is less desirable as it breaks old clients.
+
+## Changes to API servers
+
+### Calculating DiscoveryDocHashes
+
+Every API server instance calculates the DiscoveryDocHashes based on its in
+memory discovery documents for **built-in** resources. One API server instance
+only generates one DiscoveryDocHashes. And because built-in resources are
+static, the hash does not change through the life time of the API server
+process.
+
+The hash does not include the CRDs, because API server instances *can* serve
+consistently if using the `CRD` API as a coordination point. For example, when
+serving a CR request, if the API server always uses a [handler][] built based on
+the latest `CRD` stored in etcd, then all API server instances will behave
+consistently.
+
+[handler]:https://github.com/kubernetes/kubernetes/blob/63a43402a365f7a7615e01ea3e174fb8a71d67a8/staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/customresource_handler.go#L200
+
+Similarly, for aggregated APIs, the API servers can use the `APIService` API as
+a coordination point to achieve consistency.
+
+### Periodically reporting DiscoveryDocHashes & Garbage Collection
+
+Because API server instances come and go, we need a mechanism to garbage collect
+stale DiscoveryDocHashes records created by decommissioned API servers.
+
+For this purpose, every API server needs to periodically updates the
+`lastHeartbeatTime` of its record in `DiscoveryDocHashes`. Tentatively we will
+set the update period to 5 minutes. If an entry hasn't been updated for 3
+periods (15 mins), then any API server instance can remove the entry.
+
+### Handling discovery requests
+
+When responding discovery requests, apart from the original discovery documents,
+the API server also sets a `Consistent` field to indicate if all API servers are
+serving the same discovery documents, i.e., if all hash values recorded in
+`DiscoveryDocHashes` are equal. 
+
```

Comment:
i think there are some corner cases that there is a small time window that 1) it reports consistent, but it is actually not consistent 2) it reports inconsistent, but it is actually consistent. could you talk about such corner cases?

Reply:
While the later is probably less of an issue, the former is more important and it seems it's possible for some short time. Can we at least have a discussion in this doc about it?

Reply:
Discussed in the FAQ section in #1201: https://github.com/kubernetes/enhancements/pull/1201/files#diff-96f62b6e017366ceb442221838af248bR241.